From 5ceb109f555f9a2321e104336af590164bab907c Mon Sep 17 00:00:00 2001
From: Sailaja <sbandarupalli@habana.ai>
Date: Mon, 28 Nov 2022 14:13:42 +0200
Subject: [PATCH] [SW-112227] gaudi2: fail secondary image load

In ppboot, while loading preboot secondary image need to
check SEC_DIS efuse bit. If SEC_DIS bit is set, shouldn't
load secondary image

Change-Id: I4c922420f2c014be5023a071448cde6e0bbe63b0
Signed-off-by: Sailaja <sbandarupalli@habana.ai>
(cherry picked from commit ef28864a2097d6f0a8f27728abf67200592c19cc)
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index edc9c80..eb7e7e2 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -343,6 +343,13 @@
 	int ret;
 	uint32_t magic = 0;
 	uint8_t hash[SHA384_SUM_LEN] = { 0 };
+	u32 config_fuse = sp_sram_readl(SECURITY_EFUSE_SYS_CFG_OFFSET);
+
+	if ((image_ptr == ARCBOOT_SECONDARY) &&
+		(config_fuse & EFUSE_SYSCFG_DISABLE_SECONDARY_IMG)) {
+		preboot_puts("preboot secondary image disabled!\n");
+		return ARCBOOT_BRINGUP_ERROR;
+	}
 
 	flash_offset = ppboot_flash_image_offset(image_ptr);
 
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 380f833..86513ab 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -97,6 +97,7 @@
 #define EFUSE_SYSCFG_DISABLE_TWI_IF_MASK	BIT(19)
 #define EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI	BIT(8)
 #define EFUSE_SYSCFG_ENABLE_VALIDATION_MASK	BIT(10)
+#define EFUSE_SYSCFG_DISABLE_SECONDARY_IMG	BIT(21)
 
 #define RECOVERY_TEXT_BASE	(0x0 + 0x4)
 #define RECOVERY_SIZE		SZ_16K
